src/main/java/org/broadinstitute/barclay/argparser/CommandLinePluginDescriptor.java

```diff
@@ -76,103 +110,127 @@
+     * Base class for all command line plugin classes managed by this descriptor. Subclasses of
+     * this class in any of the packages returned by {@link #getPackageNames} will be command line
+     * accessible.
      */
-    public Predicate<Class<?>> getClassFilter() { return c -> true;}
+    public abstract Class<?> getPluginBaseClass();
 
     /**
-     * Return an instance of the specified pluggable class. The descriptor should
-     * instantiate or otherwise obtain (possibly by having been provided an instance
-     * through the descriptor's constructor) an instance of this plugin class.
-     * The descriptor should maintain a list of these instances so they can later
-     * be retrieved by {@link #getAllInstances}.
+     * Determine if a plugin class should be retained for this descriptor. Return true if the
+     * descriptor wants this class to be included in from the list of plugins discovered dynamically.
      *
-     * In addition, implementations should recognize and reject any attempt to instantiate
-     * a second instance of a plugin that has the same simple class name as another plugin
-     * controlled by this descriptor (which can happen if they have different qualified names
-     * within the base package used by the descriptor) since the user has no way to disambiguate
-     * these on the command line).
+     * @return false if the plugin class shouldn't be used; otherwise true
+     */
+    public boolean includePluginClass(Class<?> c) { return true;}
+
+    /**
+     * Return an instance of the specified plugin class. The descriptor should instantiate or otherwise
+     * obtain (possibly by having been provided a default instance through the descriptor's constructor)
+     * an instance of this plugin class. The descriptor should maintain a list of these instances so
+     * they can later be retrieved by {@link #getResolvedInstances}.
      *
-     * @param pluggableClass a plugin class discovered by the command line parser that
-     *                       was not rejected by {@link #getClassFilter}
+     * In addition, implementations should recognize and reject any attempt to instantiate a second
+     * instance of a plugin that has the same simple class name as another plugin controlled by this
+     * descriptor (which can happen if they have different qualified names within the base package
+     * used by the descriptor) since the user has no way to disambiguate these on the command line).
+     *
+     * @param pluginClass a plugin class discovered by the command line parser that
+     *                       was not rejected by {@link #includePluginClass}
      * @return the instantiated object that will be used by the command line parser
      * as an argument source
      * @throws IllegalAccessException
      * @throws InstantiationException
      */
-    public abstract Object getInstance(Class<?> pluggableClass)
+    public abstract T createInstanceForPlugin(Class<?> pluginClass)
```

Comment:
Why the pluginClass isn't bounded? Unless it is a factory class, it should be at least `T`, no?

Reply:
See above comment.

Reply:
It will be nice to have a javadoc comment that the param is supposed to be of class `T`, and maybe enforce to throw a `CastClassException` if the `pluginClass` is not lower-bounded.

Reply:
Updated all the javadoc to include the type bounds where appropriate.